src/hotspot/share/interpreter/linkResolver.cpp

```diff
@@ -1355,6 +1355,15 @@ void LinkResolver::runtime_resolve_virtual_method(CallInfo& result,
   // a missing receiver might result in a bogus lookup.
   assert(resolved_method->method_holder()->is_linked(), "must be linked");
 
+  // Receriver should be compatible with resolved klass, i.e. it must be a type of resolved klass
+  // or a subtype of resolved klass(receiver instanceof resolved_klass)
+  if (!recv.is_null() && !recv()->is_a(resolved_klass)) {
+    ResourceMark rm(THREAD);
+    stringStream ss;
+    ss.print("Receiver should be type/subtype of %s", resolved_klass->external_name());
```

Comment:
You should report the actual receiver type as well, as anyone encountering this exception would be completely baffled by it.

Reply:
Done. I added a jtreg test case to reproduce it.